### PR TITLE
[Improvement] Enable set ServerLoadProtection fot Master/Worker

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

- Refactor the master/ worker server overload logic.
- Enable closing the overload protection
- Support set jvmMemory/SystemMemory/cpuUsage/diskUsage protection.
- Get the metrics from actuator.

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 1]
```diff
@@ -506,10 +506,15 @@ master:
     MASTER_TASK_COMMIT_INTERVAL: "1s"
     # -- master state wheel interval, the unit is second
     MASTER_STATE_WHEEL_INTERVAL: "5s"
-    # -- Master max cpuload avg, only higher than the system cpu load average, master server can schedule
-    MASTER_MAX_CPU_LOAD_AVG: "1"
-    # -- Master reserved memory, only lower than system available memory, master server can schedule, the unit is G
-    MASTER_RESERVED_MEMORY: "0.3"
+    MASTER_SERVER_LOAD_PROTECTION_ENABLED: true
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml]: Lack of comments

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 2]
```diff
@@ -506,10 +506,15 @@ master:
...
+    # Master max cpu usage, when the master's cpu usage is smaller then this value, master server can execute workflow.
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml]:
```suggestion
    # -- Master max cpu usage, when the master's cpu usage is smaller then this value, master server can execute workflow.
```

Refer to [helm-docs](https://github.com/norwoodj/helm-docs), comments need to be added with -- for them to be recognized, then they will be automatically updated to README.md by helm-docs.

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 3]
```diff
@@ -506,10 +506,15 @@ master:
...
+    MASTER_SERVER_LOAD_PROTECTION_MAX_CPU_USAGE_PERCENTAGE_THRESHOLDS: 0.7
+    # Master max JVM memory usage , when the master's jvm memory usage is smaller then this value, master server can execute workflow.
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml]:
```suggestion
    # -- Master max JVM memory usage , when the master's jvm memory usage is smaller then this value, master server can execute workflow.
```

Similarly, comments for newly added parameters in `values.yaml` need to be modified.

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 1 above)
Comment (Gallardot) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: If deployed through k8s, this default value should be able to be set to false.

Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: Done.

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 2 above)
Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: Done

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 3 above)
Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: Done

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 4]
```diff
@@ -621,18 +627,28 @@ worker:
...
+    # If set true, will open worker overload protection
+    WORKER_SERVER_LOAD_PROTECTION_ENABLED: false
+    # -- Worker max cpu usage, when the worker's cpu usage is smaller then this value, worker server can be dispatched tasks.
+    WORKER_SERVER_LOAD_PROTECTION_MAX_CPU_USAGE_PERCENTAGE_THRESHOLDS: 0.7
+    # -- Worker max jvm memory usage , when the worker's jvm memory usage is smaller then this value, worker server can be dispatched tasks.
+    WORKER_SERVER_LOAD_PROTECTION_MAX_JVM_MEMORY_USAGE_PERCENTAGE_THRESHOLDS: 0.7
+    # -- Worker max memory usage , when the worker's memory usage is smaller then this value, worker server can be dispatched tasks.
+    WORKER_SERVER_LOAD_PROTECTION_MAX_SYSTEM_MEMORY_USAGE_PERCENTAGE_THRESHOLDS: 0.7
+    # -- Worker max disk usage , when the worker's disk usage is smaller then this value, worker server can be dispatched tasks.
+    WORKER_SERVER_LOAD_PROTECTION_MAX_DISK_USAGE_PERCENTAGE_THRESHOLDS: 0.7
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml:639]: A comment for `WORKER_EXEC_THREADS` was deleted by mistake
```
# -- Worker execute thread number to limit task instances
```

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 5]
```diff
@@ -621,18 +627,28 @@ worker:
...
     WORKER_EXEC_THREADS: "100"
     # -- Worker heartbeat interval, the unit is second
     WORKER_HEARTBEAT_INTERVAL: "10s"
-    # -- Worker heartbeat error threshold
-    WORKER_HEART_ERROR_THRESHOLD: "5"
     # -- Worker host weight to dispatch tasks
     WORKER_HOST_WEIGHT: "100"
+    # tenant corresponds to the user of the system, which is used by the worker to submit the job. If system does not have this user, it will be automatically created after the parameter worker.tenant.auto.create is true.
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml]: miss `--`

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 6]
```diff
@@ -621,18 +627,28 @@ worker:
...
+    WORKER_TENANT_CONFIG_AUTO_CREATE_TENANT_ENABLED: true
+    # Scenes to be used for distributed users. For example, users created by FreeIpa are stored in LDAP. This parameter only applies to Linux, When this parameter is true, worker.tenant.auto.create has no effect and will not automatically create tenants.
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml]: miss --

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 7]
```diff
@@ -621,18 +627,28 @@ worker:
...
+    WORKER_TENANT_CONFIG_DISTRIBUTED_TENANT: false
+    # If set true, will use worker bootstrap user as the tenant to execute task when the tenant is `default`;
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml]: miss --

File: deploy/kubernetes/dolphinscheduler/values.yaml [hunk 8]
```diff
@@ -621,18 +627,28 @@ worker:
       # -- `PersistentVolumeClaim` size
       storage: "20Gi"
   env:
-    # -- Worker max cpu load avg, only higher than the system cpu load average, worker server can be dispatched tasks
-    WORKER_MAX_CPU_LOAD_AVG: "1"
-    # -- Worker reserved memory, only lower than system available memory, worker server can be dispatched tasks, the unit is G
-    WORKER_RESERVED_MEMORY: "0.3"
-    # -- Worker execute thread number to limit task instances
+    # If set true, will open worker overload protection
```
Comment (Gallardot) [inline deploy/kubernetes/dolphinscheduler/values.yaml]: miss --

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 5 above)
Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: Done

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 4 above)
Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml:639]: Done

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 6 above)
Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: Done

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 7 above)
Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: Done

File: deploy/kubernetes/dolphinscheduler/values.yaml (hunk 8 above)
Comment (ruanwenjun) [reply deploy/kubernetes/dolphinscheduler/values.yaml]: Done

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            return true;
+        }
+        if (systemMetrics.getDiskUsedPercentage() > maxDiskUsagePercentageThresholds) {
+            log.info("Master OverLoad: the DiskUsedPercentage: {} is over then the MaxCpuUsagePercentageThresholds {}",
+                    systemMetrics.getDiskUsedPercentage(), maxCpuUsagePercentageThresholds);
+            return true;
+        }
+        if (systemMetrics.getSystemMemoryUsedPercentage() > maxSystemMemoryUsagePercentageThresholds) {
+            log.info(
+                    "Worker OverLoad: the SystemMemoryUsedPercentage: {} is over then the MaxSystemMemoryUsagePercentageThresholds {}",
```
Comment (Gallardot) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java]: typo?

Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java]: Yes, typo.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        if (systemMetrics.getTotalCpuUsedPercentage() > maxCpuUsagePercentageThresholds) {
+            log.info(
+                    "Master OverLoad: the TotalCpuUsedPercentage: {} is over then the MaxCpuUsagePercentageThresholds {}",
+                    systemMetrics.getTotalCpuUsedPercentage(), maxCpuUsagePercentageThresholds);
+            return true;
+        }
+        if (systemMetrics.getJvmMemoryUsedPercentage() > maxJVMMemoryUsagePercentageThresholds) {
+            log.info(
+                    "Master OverLoad: the JvmMemoryUsedPercentage: {} is over then the maxCpuUsagePercentageThresholds {}",
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java]:
```suggestion
                    "Master OverLoad: the JvmMemoryUsedPercentage: {} is over then the maxJVMMemoryUsagePercentageThresholds {}",
```

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+                    systemMetrics.getJvmMemoryUsedPercentage(), maxCpuUsagePercentageThresholds);
+            return true;
+        }
+        if (systemMetrics.getDiskUsedPercentage() > maxDiskUsagePercentageThresholds) {
+            log.info("Master OverLoad: the DiskUsedPercentage: {} is over then the MaxCpuUsagePercentageThresholds {}",
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java]:
```suggestion
            log.info("Master OverLoad: the DiskUsedPercentage: {} is over then the maxDiskUsagePercentageThresholds {}",
```

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+    private boolean enabled = true;
+
+    private double maxCpuUsagePercentageThresholds = 0.7;
+
+    private double maxJVMMemoryUsagePercentageThresholds = 0.7;
+
+    private double maxSystemMemoryUsagePercentageThresholds = 0.7;
+
+    private double maxDiskUsagePercentageThresholds = 0.7;
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java:41]: keep the same default value with config.
```suggestion
    private double maxCpuUsagePercentageThresholds = 0.8;

    private double maxJVMMemoryUsagePercentageThresholds = 0.8;

    private double maxSystemMemoryUsagePercentageThresholds = 0.8;

    private double maxDiskUsagePercentageThresholds = 0.8;
```

File: dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionTest.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package org.apache.dolphinscheduler.server.master.config;
+
+import org.apache.dolphinscheduler.meter.metrics.SystemMetrics;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+class MasterServerLoadProtectionTest {
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionTest.java:25]: :+1:

File: dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionTest.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Test
+    void isOverload() {
+        MasterServerLoadProtection masterServerLoadProtection = new MasterServerLoadProtection();
+        SystemMetrics systemMetrics = SystemMetrics.builder()
+                .jvmMemoryUsedPercentage(0.71)
+                .systemMemoryUsedPercentage(0.71)
+                .totalCpuUsedPercentage(0.71)
+                .diskUsedPercentage(0.71)
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionTest.java:34]:
```suggestion
                .jvmMemoryUsedPercentage(0.81)
                .systemMemoryUsedPercentage(0.81)
                .totalCpuUsedPercentage(0.81)
                .diskUsedPercentage(0.81)
```

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java (hunk 10 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java]: Done

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java (hunk 12 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java:41]: Good catch, done.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java (hunk 11 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java]: Done

File: dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionTest.java (hunk 14 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/test/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionTest.java:34]: I change the default configuration to `0.7`, so this test case doesn't need to change.

**Review by caishunfeng — APPROVED** [review]
LGTM

**Pull request merged**